actor/builtin/miner/miner_test.go

```diff
@@ -696,6 +699,115 @@ func TestMinerSubmitPoStPowerUpdates(t *testing.T) {
 	})
 }
 
+func TestMinerSubmitPoStVerification(t *testing.T) {
+	tf.UnitTest(t)
+
+	message := types.NewMessage(address.TestAddress, address.TestAddress2, 0, types.ZeroAttoFIL, "submitPoSt", nil)
+	comm1 := th.MakeCommitments()
+	comm2 := th.MakeCommitments()
+	comm3 := th.MakeCommitments()
+
+	t.Run("Sends correct parameters to post verifier", func(t *testing.T) {
+		minerState := *NewState(address.TestAddress, address.TestAddress, peer.ID(""), types.OneKiBSectorSize)
+		minerState.ProvingPeriodEnd = types.NewBlockHeight(ProvingPeriodDuration(types.OneKiBSectorSize))
+		minerState.SectorCommitments = NewSectorSet()
+		minerState.SectorCommitments.Add(1, types.Commitments{CommR: comm1.CommR})
+		minerState.SectorCommitments.Add(2, types.Commitments{CommR: comm2.CommR})
+		minerState.SectorCommitments.Add(3, types.Commitments{CommR: comm3.CommR})
+
+		// The 3 sector is not in the proving set, so its CommR should not appear in the VerifyPoSt request
+		minerState.ProvingSet = types.NewIntSet(1, 2)
+		vmctx := th.NewFakeVMContext(message, minerState)
+
+		verifier := &th.FakeVerifier{Valid: true}
+		vmctx.TestVerifier = verifier
+
+		miner := Actor{Bootstrap: false}
+
+		testProof := []types.PoStProof{th.MakeRandomPoStProofForTest()}
+		_, err := miner.SubmitPoSt(vmctx, testProof, types.EmptyIntSet())
+		require.NoError(t, err)
+
+		require.NotNil(t, verifier.PoStRequest)
+		assert.Equal(t, types.OneKiBSectorSize, verifier.PoStRequest.SectorSize)
+
+		seed := types.PoStChallengeSeed{}
+		copy(seed[:], vmctx.TestRandomness)
+
+		sortedRs := proofs.NewSortedCommRs(comm1.CommR, comm2.CommR)
+
+		assert.Equal(t, seed, verifier.PoStRequest.ChallengeSeed)
+		assert.Equal(t, 0, len(verifier.PoStRequest.Faults))
+		assert.Equal(t, 1, len(verifier.PoStRequest.Proofs))
+		assert.Equal(t, testProof, verifier.PoStRequest.Proofs)
+		assert.Equal(t, 2, len(verifier.PoStRequest.SortedCommRs.Values()))
+		assert.Equal(t, sortedRs.Values()[0], verifier.PoStRequest.SortedCommRs.Values()[0])
+		assert.Equal(t, sortedRs.Values()[1], verifier.PoStRequest.SortedCommRs.Values()[1])
+	})
+
+	t.Run("Faults if proving set commitment is missing from sector commitments", func(t *testing.T) {
+		minerState := *NewState(address.TestAddress, address.TestAddress, peer.ID(""), types.OneKiBSectorSize)
+		minerState.ProvingPeriodEnd = types.NewBlockHeight(ProvingPeriodDuration(types.OneKiBSectorSize))
+		minerState.SectorCommitments = NewSectorSet()
+
+		minerState.ProvingSet = types.NewIntSet(4)
+		vmctx := th.NewFakeVMContext(message, minerState)
+
+		miner := Actor{Bootstrap: false}
+
+		testProof := []types.PoStProof{th.MakeRandomPoStProofForTest()}
+		code, err := miner.SubmitPoSt(vmctx, testProof, types.NewIntSet())
+		require.Error(t, err)
+		assert.Equal(t, "miner ProvingSet sector id 4 missing in SectorCommitments", err.Error())
+		assert.True(t, vmerrors.IsFault(err))
+		assert.Equal(t, uint8(1), code)
+	})
```

Comment:
Just confirming my understanding: this is testing an impossible case if the actor's abstraction boundary is respected, right. This should be an invariant maintained by the actor methods and only testable because we are reaching in to modify its state, and not reproducible as a "black box" test.

I'm not sure in general that we should attempt to test things like this exhaustively, but we *should* attempt all possible ways to get the miner state to be inconsistent by manipulating its exposed methods.

(Somewhere in the VM we should be doing a `recover()` to prevent a broken actor from taking down the vm/node entirely).

Reply:
This is correct. We are deliberately forcing a violation of the invariant and checking that the code correctly reports it as a fault. I agree that it's probably not necessary to test all these cases, but the ability to do this type of testing is new, so I was trying to get a feel for what 100% coverage looks like.